server/src/main/java/org/opensearch/search/startree/StarTreeFilter.java

```diff
@@ -0,0 +1,228 @@
+                if (foundLeafNode && globalRemainingPredicateColumns == null) {
+                    globalRemainingPredicateColumns = new HashSet<>(remainingPredicateColumns);
+                }
+                currentDimensionId = dimensionId;
+            }
+
+            if (remainingPredicateColumns.isEmpty()) {
+                int docId = starTreeNode.getAggregatedDocId();
+                docIds.add(docId);
+                matchedDocsCountInStarTree++;
+                maxDocNum = Math.max(docId, maxDocNum);
+                continue;
+            }
+
+            if (starTreeNode.isLeaf()) {
+                for (long i = starTreeNode.getStartDocId(); i < starTreeNode.getEndDocId(); i++) {
+                    docIds.add((int) i);
+                    matchedDocsCountInStarTree++;
+                    maxDocNum = Math.max((int) i, maxDocNum);
+                }
+                continue;
+            }
+
+            String childDimension = dimensionNames.get(dimensionId + 1);
+            StarTreeNode starNode = null;
+            if (globalRemainingPredicateColumns == null || !globalRemainingPredicateColumns.contains(childDimension)) {
+                starNode = starTreeNode.getChildStarNode();
+            }
+
+            if (remainingPredicateColumns.contains(childDimension)) {
+                long queryValue = queryMap.get(childDimension); // Get the query value directly from the map
+                StarTreeNode matchingChild = starTreeNode.getChildForDimensionValue(queryValue);
+                if (matchingChild != null) {
+                    queue.add(matchingChild);
+                    foundLeafNode |= matchingChild.isLeaf();
+                }
+            } else {
+                if (starNode != null) {
+                    queue.add(starNode);
+                    foundLeafNode |= starNode.isLeaf();
+                } else {
+                    Iterator<? extends StarTreeNode> childrenIterator = starTreeNode.getChildrenIterator();
+                    while (childrenIterator.hasNext()) {
+                        StarTreeNode childNode = childrenIterator.next();
+                        if (childNode.getStarTreeNodeType() != StarTreeNodeType.STAR.getValue()) {
+                            queue.add(childNode);
+                            foundLeafNode |= childNode.isLeaf();
+                        }
+                    }
+                }
+            }
+        }
+
+        adder = docsWithField.grow(docIds.size());
+        for (int id : docIds) {
+            adder.add(id);
+        }
+        return new StarTreeResult(
+            docsWithField,
+            globalRemainingPredicateColumns != null ? globalRemainingPredicateColumns : Collections.emptySet(),
+            matchedDocsCountInStarTree,
+            maxDocNum
+        );
+    }
+
+    /**
+     * Helper class to wrap the result from traversing the star tree.
+     * */
+    public static class StarTreeResult {
```

Comment:
This can be `private`.

Reply:
Oh yes, mistakenly set it to public while writing tests.